# embedding Technique
converting text to vectors


In [11]:
import os
from dotenv import load_dotenv
load_dotenv()# to load environment variables from .env file

True

In [12]:
os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")  # Set your OpenAI API key

In [13]:
from langchain_openai import OpenAIEmbeddings
embeddings = OpenAIEmbeddings(model = "text-embedding-3-large")
embeddings 

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x0000011CE3D7A6B0>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x0000011CE3BCB3D0>, model='text-embedding-3-large', dimensions=None, deployment='text-embedding-ada-002', openai_api_version=None, openai_api_base=None, openai_api_type=None, openai_proxy=None, embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

In [14]:
text = "This is a sample text to be embedded."
query_result = embeddings.embed_query(text)
query_result

[-0.0016352433012798429,
 -0.02262924239039421,
 -0.027940569445490837,
 0.014017038978636265,
 0.012530242092907429,
 -0.016195805743336678,
 -0.02577115409076214,
 0.07159069180488586,
 -0.013371825218200684,
 -0.00045702653005719185,
 0.007793997414410114,
 0.037665531039237976,
 0.007256319280713797,
 -0.02371395006775856,
 0.012118800543248653,
 -8.917168088373728e-06,
 -0.02670624665915966,
 -0.0010572391329333186,
 -0.015017587691545486,
 0.0029455418698489666,
 0.046642422676086426,
 -0.016466982662677765,
 -0.01904783770442009,
 0.04148070886731148,
 0.031213393434882164,
 8.67882845341228e-05,
 -0.031306903809309006,
 0.02074035443365574,
 -0.01784156821668148,
 0.010454336181282997,
 0.009874578565359116,
 0.012838821858167648,
 0.006021996960043907,
 -0.028688643127679825,
 0.033476315438747406,
 0.00601732125505805,
 0.024985676631331444,
 0.05494604632258415,
 -0.0008053485653363168,
 -0.020759057253599167,
 0.04611876979470253,
 -0.017682602629065514,
 -0.023545634001493

In [15]:
len(query_result)  # Length of the embedding vector

3072

In [16]:
query_result[0]

-0.0016352433012798429

I can also set my own dimension, unlike above which is 3072 etc

In [17]:
embeddings_1024 = OpenAIEmbeddings(model = "text-embedding-3-large", dimensions=1024)
query_result_1024 = embeddings_1024.embed_query(text)
query_result_1024

[-0.0022603871766477823,
 -0.031280267983675,
 -0.03862208500504494,
 0.019375670701265335,
 0.017320480197668076,
 -0.022387366741895676,
 -0.03562331572175026,
 0.09895940124988556,
 -0.01848379522562027,
 -0.000631745089776814,
 0.010773597285151482,
 0.052064843475818634,
 0.010030367411673069,
 -0.03277965262532234,
 0.016751747578382492,
 -1.2326148862484843e-05,
 -0.03691588714718819,
 -0.0014614154351875186,
 -0.020758723840117455,
 0.004071605391800404,
 0.06447354704141617,
 -0.02276221290230751,
 -0.02632971480488777,
 0.057338543236255646,
 0.04314609244465828,
 0.00011996694229310378,
 -0.04327534884214401,
 0.028669271618127823,
 -0.02466229535639286,
 0.01445096731185913,
 0.013649571686983109,
 0.017747029662132263,
 0.008324170485138893,
 -0.0396561436355114,
 0.04627411812543869,
 0.008317708037793636,
 0.034537553787231445,
 0.07595159858465195,
 -0.0011132286163046956,
 -0.028695123270154,
 0.06374970823526382,
 -0.024442557245492935,
 -0.0325469896197319,
 0.026187

In [18]:
len(query_result_1024)

1024

In [19]:
query_result_1024[0]

-0.0022603871766477823

# now in this step we will do as following
use textloader

use RecurssiveTextSplitter

In [20]:
from langchain_community.document_loaders import TextLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

loader = TextLoader(r'speeh.txt')
docs = loader.load()

docs


[Document(metadata={'source': 'speeh.txt'}, page_content='I have a dream that one day every valley shall be exalted, every hill and mountain shall be made low, the rough places will be made plain, and the crooked places will be made straight; and the glory of the Lord shall be revealed, and all flesh shall see it together.\n\nThis is our hope. This is the faith that I go back to the South with.\n\nWith this faith, we will be able to hew out of the mountain of despair a stone of hope.')]

In [21]:
text_splitter = RecursiveCharacterTextSplitter(chunk_size = 30, chunk_overlap=3)
final_docs=text_splitter.split_documents(docs)
final_docs

[Document(metadata={'source': 'speeh.txt'}, page_content='I have a dream that one day'),
 Document(metadata={'source': 'speeh.txt'}, page_content='every valley shall be'),
 Document(metadata={'source': 'speeh.txt'}, page_content='be exalted, every hill and'),
 Document(metadata={'source': 'speeh.txt'}, page_content='mountain shall be made low,'),
 Document(metadata={'source': 'speeh.txt'}, page_content='the rough places will be made'),
 Document(metadata={'source': 'speeh.txt'}, page_content='plain, and the crooked places'),
 Document(metadata={'source': 'speeh.txt'}, page_content='will be made straight; and'),
 Document(metadata={'source': 'speeh.txt'}, page_content='the glory of the Lord shall'),
 Document(metadata={'source': 'speeh.txt'}, page_content='be revealed, and all flesh'),
 Document(metadata={'source': 'speeh.txt'}, page_content='shall see it together.'),
 Document(metadata={'source': 'speeh.txt'}, page_content='This is our hope. This is the'),
 Document(metadata={'source':

# Vector ebedding and Vector DB (for storing)

In [22]:
from langchain_community.vectorstores import Chroma

db = Chroma.from_documents(final_docs, embeddings_1024)
# 1st argument is the splitted documents, 2nd is the embeddings
db

# how to query from this vector DB

In [23]:
# so using the following text 
# # we are finding the similarity of it from the entire text 
# that is in the vector DB and we are getting the most similar text to the query text

query = "This is our hope. This is the faith that I go back to the South with."
retrieved_results = db.similarity_search(query)
print(retrieved_results)

[Document(metadata={'source': 'speeh.txt'}, page_content='This is our hope. This is the'), Document(metadata={'source': 'speeh.txt'}, page_content='With this faith, we will be'), Document(metadata={'source': 'speeh.txt'}, page_content='faith that I go back to the'), Document(metadata={'source': 'speeh.txt'}, page_content='of hope.')]
